### ISSUE TITLE
Split middleware

### ISSUE THREAD

**Issue by skalee** [riboseinc/rack-cleanser#8]
AFAIU, the Rack Cleanser does two things:

* halts request processing and renders nice 404 when given request contains unexpected characters
* validates parameter length

IMO these two things are pretty distinct. Also, I've noticed either can be turned off, depending on configuration.

I would like to split this middleware into two smaller ones, each doing its jobs and nothing more. I believe this way it will be simpler to use and to configure. Both rack apps can reside in the same repository, and if desired, even in the same gem. Can I?

cc: @ronaldtse @ribose-jeffreylau.

**Comment by ronaldtse** [comment]
Yes @skalee , indeed both do quite different things but are part of a set of checks that perform security-related validations.

We found inspiration having both in the same Rack app from https://github.com/kickstarter/rack-attack, and the benefit of having one Rack app is to not enclose the entire stack with one more level.

We should have more "cleansing" functions coming soon so it might be better to keep them together, and we can implement a DSL like https://github.com/kickstarter/rack-attack ?

```ruby
class Rack::Attack
  Rack::Attack.cache.store = ActiveSupport::Cache::MemoryStore.new 

  throttle('req/ip', :limit => 300, :period => 5.minutes) do |req|
    req.ip # unless req.path.start_with?('/assets')
  end

  # Key: "rack::attack:#{Time.now.to_i/:period}:logins/ip:#{req.ip}"
  throttle('logins/ip', :limit => 5, :period => 20.seconds) do |req|
    if req.path == '/login' && req.post?
      req.ip
    end
  end
```

**Comment by skalee** [comment]
I've just generated a fresh Rails application and its middleware stack consists of over 20 items.  I don't think that adding another dozen of Rack Cleanser's middlewares would make a dramatic difference.  Visible — yes, but not dramatic. However I agree that a single middleware approach has its advantages too.

On one hand, splitting the Rack Cleanser would make configuration really straightforward without creating a dedicated DSL or things like that.  Moreover, not only it would allow developers to cherry-pick functionalities they want with ease, but also to reorder them, or even to place their own middleware between these provided by us.  And such thing would be difficult to achieve with a single middleware approach.  On the other hand, if a single middleware controls cleansers execution, then we manage the order in which they are executed, and we can make it optimal for most (if not all) users.  Also, I guess that it also makes gem upgrades slightly easier for its users. 
 When it comes to rack-attack-alike configuration, I don't expect many difficulties with either approach, so it's really only a matter of whether we want to give users more control or not.

Whatever approach we take, I am going to refactor code anyway so that particular cleansers are better separated, and the `Cleanser` class does really nothing but executing cleansers one by one.  Configuration method calls (like `limit_param_length(name, options = {}, &block)`) will be delegated to proper cleanser subclasses.

**Comment by ronaldtse** [comment]
@skalee I agree that separated middlewares would allow the user to potentially have more control over their order and how they can best integrate them into their own middleware stack.

However today most people suffer from too many choices, and probably most users (including us) would prefer keeping the Cleanser by itself since splitting it would make the stack traces deeper without real need.

In any case, even if we keep with the single Cleanser middleware, a user can always instantiate two copies (or more) of them to selectively enable functionality 1 / 2 respectively and insert them in the right places.
